src/DiffSharp.Core/Tensor.fs

```diff
@@ -42,7 +42,13 @@ type Tensor =
         | TensorF(_) -> failwith "Cannot cast TensorF - do not cast during differentiation"
         | TensorR(_) -> failwith "Cannot cast TensorR - do not cast during differentiation"
 
-    member t.move(backend) =
+    member t.move(backend: Backend) =
+        // If a backend move is needed then first move to the CPU
+        let t = 

```

Comment:
@gbaydin Just to explain one things here

Initially in this PR I adjusted the backend to mimic supporting both GPU and CPU.  This would allow us to better pin down the expected behaviour for things like moving between devices and backends.  However it doubles tests times so I took it out

However that did catch a case where the DataLoader move test was moving data from "Reference GPU" to "Torch CPU".  

In cases of a backend move I think we should always move to CPU first, and then move backends, and then to the target device in the other backend.